staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,420 @@
+	// namespace/names known to exist for that definition
+	definitionInfo map[string]definitionInfo
+
+	bindingInfos map[string]bindingInfo
+
+	mutex sync.Mutex
+}
+
+type definitionInfo struct {
+	paramGr           schema.GroupVersionKind
+	dependentBindings sets.String
+}
+
+type bindingInfo struct {
+	definitionName string
+
+	// Compiled CEL expressoin turned into an evaluator
+	evaluator EvaluatorFunc
+}
+
+type paramInfo struct {
+	controller           generic.Controller[*unstructured.Unstructured]
+	stop                 func()
+	dependentDefinitions sets.String
+}
+
+func NewAdmissionController(
+	//
+	policyDefinitionsInformer cache.SharedIndexInformer,
+	policyBindingInformer cache.SharedIndexInformer,
+
+	// Fetches OpenAPI specifications for a given GroupVersionResource
+	//
+	openapiFetcher func(schema.GroupVersionResource) interface{},
+	restMapper meta.RESTMapper,
+	dynamicClient dynamic.Interface,
+) *celAdmissionController {
+	c := &celAdmissionController{
+		definitionInfo:       make(map[string]definitionInfo),
+		bindingInfos:         make(map[string]bindingInfo),
+		paramsCRDControllers: make(map[schema.GroupVersionKind]paramInfo),
+		dynamicClient:        dynamicClient,
+		openapiFetcher:       openapiFetcher,
+		restMapper:           restMapper,
+	}
+
+	c.policyDefinitionsController = generic.New(
+		generic.NewInformer[fantasyInternalPolicyDefinition](policyDefinitionsInformer),
+		c.reconcilePolicyDefinition,
+		generic.ControllerOptions{
+			Workers: 1,
+			Name:    "cel-policy-definitions",
+		},
+	)
+	c.policyBindingController = generic.New(
+		generic.NewInformer[fantasyInternalPolicyBinding](policyBindingInformer),
+		c.reconcilePolicyBinding,
+		generic.ControllerOptions{
+			Workers: 1,
+			Name:    "cel-policy-bindings",
+		},
+	)
+	return c
+}
+
+func (c *celAdmissionController) Run(stopCh <-chan struct{}) {
+	if c.runningContext != nil {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c.runningContext = ctx
+	defer func() {
+		c.runningContext = nil
+	}()
+
+	wg := sync.WaitGroup{}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.policyDefinitionsController.Run(ctx)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.policyBindingController.Run(ctx)
+	}()
+
+	<-stopCh
+	cancel()
+	wg.Wait()
+}
+
+func (c *celAdmissionController) Validate(
+	ctx context.Context,
+	a admission.Attributes,
+	o admission.ObjectInterfaces,
+) (err error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Ask policy definitions informer to list all definitoins
+	allDefinitions, err := c.policyDefinitionsController.
+		Informer().
+		List(labels.Everything())
+
+	if err != nil {
+		//!TODO: wrap with some sort of internal error?
+		utilruntime.HandleError(err)
+		return err
+	}
+
+	var allDecisions []PolicyDecision = nil
+
+	for _, definition := range allDefinitions {
+		if !definition.Matches(a) {
+			// Binding does not match constraints
```

Comment:
This should be the policy definition match?

Reply:
It is, just a typo with comment. fixed locally to push soon